File: apps/web/lib/embeddings.ts
import OpenAI from 'openai';
import {
  EMBED_MODEL,
  EMBED_DIM,
  prepareEmbeddingInputs,
  stripMarkdownForEmbedding,
} from '@chemclaw2/agent-tools';

let client: OpenAI | undefined;

function getClient() {
  if (!client) {
    const apiKey = process.env.OPENAI_API_KEY;
    if (!apiKey) throw new Error('OPENAI_API_KEY is required for embeddings');
    client = new OpenAI({ apiKey });
  }
  return client;
}

export async function embedText(text: string): Promise<number[]> {
  const [embedding] = await embedTexts([stripMarkdownForEmbedding(text)]);
  return embedding;
}

/**
 * Batch embed multiple texts. Markdown markup is stripped before the call so
 * the model doesn't waste tokens on `**`/`#`/backticks etc. Returns vectors
 * in the same order and length as the input — empty/whitespace inputs throw
 * rather than silently desync the index alignment.
 *
 * Wave-1 D4: cap each upstream request at MAX_BATCH chunks. A 500k-char
 * wiki page yields ~400 chunks; without the cap a single request approaches
 * the OpenAI per-call limit.
 *
 * Wave-3h perf: batches now fire in parallel via Promise.all. Order is
 * preserved by indexing batches into their slots up-front. Previously the
 * sequential await chain held the wiki upsert connection open for 4 round-
 * trips on a 400-chunk page; parallel cuts that to one round-trip's latency.
 */
const MAX_BATCH = 100;

export async function embedTexts(texts: string[]): Promise<number[][]> {
  if (texts.length === 0) return [];
  const stripped = texts.map(stripMarkdownForEmbedding);
  const inputs = prepareEmbeddingInputs(stripped);

  const batches: string[][] = [];
  for (let start = 0; start < inputs.length; start += MAX_BATCH) {
    batches.push(inputs.slice(start, start + MAX_BATCH));
  }
  const client = getClient();
  const responses = await Promise.all(
    batches.map((batch) => client.embeddings.create({ model: EMBED_MODEL, input: batch })),
  );

  const out: number[][] = [];
  for (let i = 0; i < batches.length; i++) {
    const res = responses[i];
    const batch = batches[i];
    if (res.data.length !== batch.length) {
      throw new Error(`embedTexts: model returned ${res.data.length} vectors for ${batch.length} inputs`);
    }
    for (const d of res.data) {
      if (d.embedding.length !== EMBED_DIM) {
        throw new Error(`embedTexts: vector dim ${d.embedding.length} ≠ expected ${EMBED_DIM}`);
      }
      out.push(d.embedding);
    }
  }
  return out;
}

File: api/embeddings.py
"""OpenAI text-embedding-3-small client.

Shared by api/routes/wiki.py, api/agent/tools.py (semantic lookup), and
api/workers/campaign_worker.py (campaign-wiki upsert). Each module previously
imported `embed_texts` from `api/routes/wiki.py`, which forced a route module
to be importable from a worker process — moving it here breaks that cycle.
"""
from __future__ import annotations

import os

from openai import AsyncOpenAI

EMBED_MODEL = "text-embedding-3-small"
EMBED_DIM = 1536  # text-embedding-3-small returns 1536-dim vectors

_client: AsyncOpenAI | None = None


def _get_client() -> AsyncOpenAI:
    global _client
    if _client is None:
        # Explicit timeout: the SDK default is 600 s, which would let one hung
        # embedding call stall a wiki upsert or document ingest for 10 minutes.
        # max_retries=3 keeps the SDK's built-in exponential backoff for
        # transient 429/5xx (default is 2).
        _client = AsyncOpenAI(
            api_key=os.environ.get("OPENAI_API_KEY", ""),
            base_url=os.environ.get("OPENAI_BASE_URL") or None,
            timeout=30.0,
            max_retries=3,
        )
    return _client


async def embed_texts(texts: list[str]) -> list[list[float]]:
    resp = await _get_client().embeddings.create(model=EMBED_MODEL, input=texts)
    return [d.embedding for d in resp.data]
